core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala

```diff
@@ -315,130 +316,137 @@ trait SubmissionMonitor extends FutureSupport with LazyLogging with RawlsInstrum
+        _ <- processMultipleWorkflows(yesOutputs, "workflowsWithOutputs")
+        // finally, after processing all workflows, check if the submission as a whole
+        // can be completed
+        statusCheckComplete <- datasource.inTransaction { dataAccess =>
+          // update submission after workflows are updated
+          updateSubmissionStatus(dataAccess) map { shouldStop: Boolean =>
+            // return a message about whether our submission is done entirely
+            StatusCheckComplete(shouldStop)
           }
         }
-      }
+      } yield statusCheckComplete
     }
 
-  /**
-   * Batch workflowsWithOutputs into groups so as not to exceed config.attributeUpdatesPerWorkflow attributes per transaction
-   *
-   * @param workflowsWithOutputs
-   * @return
-   */
-  def batchWorkflowsWithOutputs(
-    workflowsWithOutputs: Seq[(WorkflowRecord, ExecutionServiceOutputs)]
-  ): List[Seq[(WorkflowRecord, ExecutionServiceOutputs)]] = {
-    // batch workfowsWithOutpus into groups so as not to exceed config.attributeUpdatesPerWorkflow attributes per transaction
-    val countOfAllAttributes = workflowsWithOutputs.foldLeft(0) { case (subTotal, (_, outputs)) =>
-      attributeCount(outputs.outputs.values.collect { case Left(output) => output }) + subTotal
+  private def markWorkflowFailed(workflowRecord: WorkflowRecord, fatal: RawlsFatalExceptionWithErrorReport) = {
+    logger.error(
+      s"Marking workflow ${externalId(workflowRecord)} as failed handling outputs in $submissionId with user-visible reason ${fatal.toString})"
+    )
+    datasource.inTransaction { dataAccess =>
+      dataAccess.workflowQuery.updateStatus(workflowRecord, WorkflowStatuses.Failed) andThen
+        dataAccess.workflowQuery.saveMessages(Seq(AttributeString(fatal.toString)), workflowRecord.id)
     }
-    // plus 1 because integer division rounds down
-    val batchCount = countOfAllAttributes / config.attributeUpdatesPerWorkflow + 1
-    workflowsWithOutputs.grouped(workflowsWithOutputs.size / batchCount + 1).toList
   }
 
+  private def externalId(workflowRecord: WorkflowRecord): String = workflowRecord.externalId.getOrElse("?")
+
+  private def processWorkflow(workflowRec: WorkflowRecord,
+                              execServiceOutputsOption: Option[ExecutionServiceOutputs],
+                              tracingContext: RawlsTracingContext
+  )(implicit
+    executionContext: ExecutionContext
+  ) =
+    // In an independent transaction,
+    //  - attach the outputs for this workflow, if any outputs exist
+    //  - update the status of the workflow in Rawls to match what Cromwell says
+    // If attaching outputs fails for legit reasons (e.g. they're missing), it will mark the workflow as failed. This is correct.
+    // If attaching outputs throws an exception (because e.g. deadlock or ConcurrentModificationException), the status will remain un-updated
+    // and will be re-processed next time we call queryForWorkflowStatus().
```

Comment:
These are great comments (all of them in this PR)

I have only a passing familiarity with Scala, but am trying to build context by reviewing: would you mind pointing out where these exceptions and subsequent re-processing are handled?

It sounds like deadlocks and `ConcurrentModificationException` are expected to happen occasionally, is that right? Would we be able to tell if a workflow were stuck in a reprocessing loop somehow? Is that a possibility?

Reply:
Note that this comment is repurposed from the old line 361!

> pointing out where these exceptions and subsequent re-processing are handled

This gets into the higher-level flow of this actor. The best place to start for that is the `receive` function at line 134, which shows all the things this actor is capable of. It starts with a `StartMonitorPass` message, which will find all workflows that are still running in Rawls but are completed in Cromwell. It will then pass those workflows as the payload in a `ExecutionServiceStatusResponse` message back to self. So, if a workflow hit a non-fatal exception here, it would not be marked as complete in Rawls, and the next time the actor received a `StartMonitorPass`, it would notice that workflow is running in Rawls but complete in Cromwell and would then send it off for re-processing.

This PR doesn't do anything to help with _noticing_ reprocessing loops (though I hope it helps _avoid_ a class of them!). See https://github.com/broadinstitute/rawls/pull/2292 for work on that front.

Reply:
Thank you!